spec/PeriodicBackgroundSync-index.bs

```diff
@@ -92,3 +94,60 @@ Fetch requests within the <code>periodicsync</code> event while [=in the backgro
 
 ## History Leaking ## {#history-leaking}
 Fetch requests within the <code>periodicsync</code> event while [=in the background=] may reveal something about the client's navigation history to middleboxes on networks different from the one used to create the [=periodicsync registration=]. For instance, the client might visit site https://example.com, which registers a <code>periodicsync</code> event, but based on the implementation, might not fire until after the user has navigated away from the page and changed networks. Middleboxes on the new network may see the fetch requests that the <code>periodicsync</code> event makes. The fetch requests are HTTPS so the request contents will not be leaked but the domain may be (via DNS lookups and IP address of the request). To prevent this leakage of browsing history, the user agent MAY choose to only fire <code>periodicsync</code> events on the network the [=periodicsync registration=] was made on, with the understanding that it will reduce usability by not allowing synchronization opportunistically.
+
+API Description {#api-description}
+===================================
+## Extensions to the {{ServiceWorkerGlobalScope}} interface ## {#serviceworkerglobalscope-extension}
+<script type="idl">
+partial interface ServiceWorkerGlobalScope {
+    attribute EventHandler onperiodicsync;
+};
+</script>
+
+## Extensions to the {{ServiceWorkerRegistration}} interface## {#service-worker-extensions}
+
+<script type="idl">
+[Exposed=(Window,Worker)]
+partial interface ServiceWorkerRegistration {
+  readonly attribute PeriodicSyncManager periodicSync;
+};
+</script>
+
+A {{ServiceWorkerRegistration}} has a <dfn>periodic sync manager</dfn> (a {{PeriodicSyncManager}}), initially a new {{PeriodicSyncManager}} whose [=service worker registration=] is the [=context object=]'s [=service worker registration=]. The <code>periodicSync</code> attribute's getter must return the [=context object=]'s [=periodic sync manager=].
+
+## {{PeriodicSyncManager}} interface ## {#periodicsyncmanager-interface}
+<script type="idl">
+[Exposed=(Window,Worker)]
+interface PeriodicSyncManager {
+    Promise<void> register(DOMString tag, optional BackgroundSyncOptions options);
+    Promise<sequence<DOMString>> getTags();
+    Promise<void> unregister(DOMString tag);
+};
+
+dictionary BackgroundSyncOptions {
+    [EnforceRange] unsigned long long minInterval = 0;
+};
+</script>
+
+The <code><dfn method for=PeriodicSyncManager title="register(tag, options)">register(<var>tag</var>, <var>options</var>)</dfn></code> method, when invoked, MUST return <a>a new promise</a> <var>promise</var> and run the following steps <a>in parallel</a>:
+
+1. Let <var>serviceWorkerRegistration</var> be the {{PeriodicSyncManager}}'s associated [=service worker registration=].
+1. If |serviceWorkerRegistration|’s [=active worker=] is null, [=reject=] |promise| with an {{InvalidStateError}} and abort these steps.
```

Comment:
Implementation note: Chrome seems to violate this and uses an `UnknownError` if the browser process decides that there is no live service worker.

Reply:
https://cs.chromium.org/chromium/src/third_party/blink/renderer/modules/background_sync/sync_manager.cc?rcl=11875b366c253079d5f733267325e5c85d2907a4&l=28

Reply:
https://cs.chromium.org/chromium/src/third_party/blink/renderer/modules/background_sync/periodic_sync_manager.cc?rcl=37594cbfa3400a04d9fc798cf6ecdaca0026bf6f&l=29

Reply:
That's the renderer process check, I mentioned the browser process check ;)

https://cs.chromium.org/chromium/src/third_party/blink/renderer/modules/background_sync/periodic_sync_manager.cc?g=0&l=125

Reply:
Ok, I'll take a bug to update this in the implementation for both one-shot and periodic Sync. However in this algorithm, this check is happening early on, which matches with current behavior.

Reply:
fixed now.